### PR TITLE
ekump/adding-test-agent-to-trace-utils

### PR DESCRIPTION
**Pull request by ekump**
# What does this PR do?

Introduces the test agent to libdatadog to be used for integration tests in trace-utils.

The test agent runs inside a docker container that is spun up/down within a given test via [testcontainers-rs](https://github.com/testcontainers/testcontainers-rs). 

The `DatadogTestAgent` struct has been created as a wrapper around the testcontainers implementation. It also provides some helper functions to make it easier to write tests that leverage the test agent.

A single integration test has been added that performs a snapshot test via the test-agent for a trace. 

# Motivation

There is imminent work around trace_utils planned that could potentially modify trace payloads. These tests should provide some confidence for that work.

# Additional Notes

The new tests naturally require docker. The README has been updated with instructions on how to skip these tests locally if you're not running docker. Alternatively, if people feel strongly we can invert this and default to not running these tests by default locally via a cargo feature.

The test will run on github CI only on Ubuntu for the time being.

# How to test the change?

`cargo nextest run`

### CODE REVIEW

**Review by bantonsson — APPROVED** [review]
I, for one, welcome our new docker running overlords...

File: trace-utils/src/test_utils/datadog_test_agent.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            container: DatadogTestAgentContainer::new(relative_snapshot_path)
+                .start()
+                .await
+                .expect("Unable to start DatadogTestAgent"),
+        }
+    }
+
+    async fn get_base_uri_string(&self) -> String {
+        let container_host = self.container.get_host().await.unwrap().to_string();
+        let container_port = self.container.get_host_port_ipv4(8126).await.unwrap();
```
Comment (VianneyRuhlmann) [inline trace-utils/src/test_utils/datadog_test_agent.rs:174]: Should this use `TEST_AGENT_PORT` ?

File: trace-utils/tests/snapshots/compare_v04_trace_snapshot_test.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        "origin": "cloudfunction",
+        "functionname": "dummy_function_name"
+      },
+      "metrics": {
+        "_top_level": 1.0
+      },
+      "type": "web"
+    }
+  ]
+]
```
Comment (VianneyRuhlmann) [inline trace-utils/tests/snapshots/compare_v04_trace_snapshot_test.json:63]: Very nitpick: no newline at end of file

**Review by VianneyRuhlmann — APPROVED** [review]
I'm impressed by how smooth it is to run test with docker within cargo 🤩

File: trace-utils/src/test_utils/datadog_test_agent.rs (hunk 1 above)
Comment (ekump) [reply trace-utils/src/test_utils/datadog_test_agent.rs:174]: yes, good catch.

File: trace-utils/Cargo.toml [hunk 3]
```diff
@@ -27,11 +27,13 @@ bytes = "1.6.0"
...
+cargo_metadata = { version = "0.18.1", optional = true }
 
 [dev-dependencies]
 tokio = { version = "1", features = ["macros", "rt-multi-thread"] }
 serde_json = "1.0"
 httpmock = { version = "0.7.0"}
 
 [features]
-test-utils = ["httpmock"]
+test-utils = ["httpmock", "testcontainers", "cargo_metadata"]
```
[the remark this answers is not in the document]
Comment (ekump) [reply trace-utils/Cargo.toml:39]: bleargh, my IDE was incorrectly defaulting to adding new lines. Will fix.

**Pull request merged**